### PR TITLE
Add nn header and reference for cppcli generator

### PR DESCRIPTION
**Pull request by FanglinIfolor**
* Fixed: add `#include` for nn header in C++/CLI hpp files when `--cpp-nn-header` is specified
* Fixed: add `ref class Foo;` in C++/CLI hpp files when there are one interface depends on another interface
* Added: tests for C++/CLI with nn header and referencing abstract classes
* Fixed: add `.bat` extension to binary when running `sbt assembly` on Windows
* Fixed: able to run tests on Windows
* Fixed: handle different end of lines (`\r\n` v.s. `\n`) and file path separator (`\` and `/`) when comparing strings on Windows

### CODE REVIEW

**Comment by a4z** [comment]
Awesome PR, thanks a lot!

I can not check all details, since the Windows C# part is not something I use , but on a first glance, it seems looking fine to me.

For an approval I would like to wait for @jothepro for an in depth review, since he actually uses the C# port

File: src/it/scala/djinni/GeneratorIntegrationTest.scala [hunk 1]
```diff
@@ -343,6 +343,66 @@ class GeneratorIntegrationTest extends IntegrationTest with GivenWhenThen {
       assertFileContentEquals(idlFile, "", List("generated-files.txt"))
     }
 
+    it(
+      "should be able to generate C++ and C++/CLI outputs for interfaces with external dependencies and non-null pointers"
```
Comment (jothepro) [inline src/it/scala/djinni/GeneratorIntegrationTest.scala:347]: I'm not 100% sure but from my understanding the support for non-null pointers that you added and the problem with external dependencies are two separate topics, that could be tested in isolation from each other? Am I correct with that? If so, can you split the test cases so only one feature is tested at a time?

**Review by jothepro — COMMENTED** [review]
Nice, thx for your attention to detail in this PR, including elaborate tests and updating the docs! ❤️ 

For future contributions pls consider creating multiple separate PRs for the things that you fixed, as far as they are separable. This makes it easier for me as a reviewer to catch what's going on.

> Fixed issues of missing ref class ...; in C++/CLI hpp files when referencing abstract classes

Can you give me some more insights on what the problem is and how you fixed it. I don't understand it yet.

File: src/it/scala/djinni/IntegrationTest.scala [hunk 2]
```diff
@@ -170,6 +172,19 @@ class IntegrationTest extends AnyFunSpec {
...
+    *   the string generated on Windows to be transformed
+    * @return
+    *   the string that \r is removed and \\ is replaced by / from str
+    */
+  def windowsToLinuxStringFormat(
+      str: String
+  ): String = {
+    str.replace("\r", "").replace("\\", "/")
+  }
+
```
[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala:187]: This issue only exists on Windows, and the test was not fully working on Windows before 😅.

I also don't think this approach is very elegant. Splitting lines to compare line by line sounds a good idea; I found `System.lineSeparator` could be useful here.

Comparing paths in strings could be slightly more challenging, as we will need to first find out where the path is in each line.

I will have a look again tomorrow and see if I can come up with something more elegant.

File: src/main/scala/djinni/CppMarshal.scala [hunk 3]
```diff
@@ -122,7 +122,7 @@ class CppMarshal(spec: Spec) extends Marshal(spec) {
             List(ImportRef("<memory>"))
           }
           spec.cppNnHeader match {
-            case Some(nnHdr) => ImportRef(nnHdr) :: base
+            case Some(nnHdr) => ImportRef(q(nnHdr.replace("\"", ""))) :: base
```
[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/main/scala/djinni/CppMarshal.scala:125]: I changed this because I couldn't find a way to generate exactly `#include "nn.hpp"`, given [this line of code](https://github.com/cross-language-cpp/djinni-generator/blob/main/src%2Fit%2Fscala%2Fdjinni%2FIntegrationTest.scala#L146)
```
     if (useNNHeader) {
>      cmd += " --cpp-nn-header nn.hpp"
       cmd += " --cpp-nn-type dropbox::oxygen::nn_shared_ptr"
       cmd += " --cpp-nn-check-expression NN_CHECK_ASSERT"
     }
```

I remembered when I tried to change that line to `cmd += """ --cpp-nn-header "nn.hpp""""` or many alternatives (e.g., `\"`), it would generate something like `#include ""nn.hpp""`. And I remembered when I managed to make it work in the tests, it doesn't work when I use the compiled djinni.bat to generate my real code; but it could be something to do with the batch script. I need to check again the details of the issue tomorrow.

So the work around is, we don't depend on whether user passes also double quotes in `--cpp-nn-header` or not, but we first remove all the double quotes, then add double quotes back. Then it works in both tests and my code.

Probably you have better ideas what could be the issue of generating `#include ""nn.hpp""`, then we can find more elegant way to solve the issue.

File: src/it/scala/djinni/GeneratorIntegrationTest.scala (hunk 1 above)
Comment (FanglinIfolor) [reply src/it/scala/djinni/GeneratorIntegrationTest.scala:347]: You are right. I added this test accidentally, as I wanted to add a test to test interface "dependencies" (exactly [the next test](https://github.com/cross-language-cpp/djinni-generator/pull/128/files/d9616f9dfe51483d15718327a74b02febadf5205#diff-dc8bedef9ea7fe6b135c617767cddccc31ace3e8f42f86e164d2717a63a48b74R375)), but I took the external dependencies and realized that's not what I needed.

I will split test cases and test external dependencies in one test and non-null pointers in another.

File: src/main/scala/djinni/CppMarshal.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (a4z) [reply src/main/scala/djinni/CppMarshal.scala:125]: That's a good catch! Seems there is a general problem 

There is a  `def q(s: String) = '"' + s + '"'` in the generator

Very often `ImportRef` is used like this : `ImportRef(q(d.name + ".hpp"))`  but not always ... and this might be the problem,  I guess that is not the most easy to read code in the code base, and it makes also not super clear how it should be used, but that's how it is and we got it 🙈  

maybe that line should be changed to 
```scala
spec.cppNnHeader match {
  case Some(nnHdr) => ImportRef(q(nnHdr)) :: base
```
and then 
`cmd += " --cpp-nn-header nn.hpp"` 
works 
and if we 🙏 , and have luck, nothing breaks 😁  

Can you please give that a try, and if it breaks something, tell us what, so we can unify ... (but it should work, the line is only handling (should) the cmd line parameter, just wonder how it worked so far, or was it untested ... 🤔

**Comment by FanglinIfolor** [comment]
@jothepro Thank you for the review and the suggestion regarding separating PRs, will do so in the future contribution.

> Fixed issues of missing ref class ...; in C++/CLI hpp files when referencing abstract classes
> Can you give me some more insights on what the problem is and how you fixed it. I don't understand it yet.

Here I meant to add the forward declaration `ref class Foo;` at the beginning of the code if one abstract interface uses another abstract interface `Foo`. It's not an issue if there are no circular dependencies, but when there are such dependencies, we need to use forward declaration, i.e. add `ref class Foo;`, to solve the compile error.

Note that I borrow the implementation from C++, that also generates `class Foo;` if `Foo` is an abstract class, no matter whether there are circular dependencies or not. It's possible to determine if circular dependencies exist by traversing all dependencies, but it's out of scope of this PR.

In short, I will rephrase the commit to make it clear 😅.

File: build.sbt [hunk 4]
```diff
@@ -3,6 +3,7 @@ import sbtassembly.AssemblyPlugin.defaultUniversalScript
 ThisBuild / scalaVersion := "2.13.6"
 ThisBuild / organization := "com.github.cross-language-cpp"
 
+val binExt = if (System.getProperty("os.name").startsWith("Windows")) ".bat" else ""
```
[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply build.sbt:6]: Actually, we don't need to remove that line, as in the build automation, the generator is built on Ubuntu, and it will not add `.bat` file extension in that case, so we still need to copy the `target/bin/djinni` to `target/bin/djinni.bat`, so later we can upload `target/bin/djinni.bat` as the release asset for Windows. If you agree, then I will not change anything further.

[the remark this answers is not in the document]
Comment (jothepro) [reply build.sbt:6]: Ahh, right. Sry for the confusion...

File: src/main/scala/djinni/CppMarshal.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (a4z) [reply src/main/scala/djinni/CppMarshal.scala:125]: I feel ignored, but that's OK

logically speaking, one of both must have the right quoting
either `ImportRef(q(nnHdr))` or `ImportRef(nnHdr)`

if this is not the case, then there is something special in the input (the input has been q'ed in one case but not an other) , and if there is something special in the input, it might be interesting to know where this comes from, since then this might also happen on other places. Not? 
And if, we might need to change the q function (optimally also rename but this is a different story) to ensure something is `"` as it should, and not to much or to less

Since we do no needed that special handling anywhere else, I would be really interested in the reasons

[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/main/scala/djinni/CppMarshal.scala:125]: @a4z eventually, I realized several issues:

1. The script I used to generate the code looks like: `--cpp-nn-header "nn.hpp"` which is with double quote. We needed to add double quotes because the generator doesn't add them automatically:
    ```
    spec.cppNnHeader match {
      case Some(nnHdr) => ImportRef(nnHdr) :: base
    ```
2. Current tests do not cover the case of generating `#include "nn.hpp"`. In the `main` branch, although a test set `useNNHeader` to true (see [this line](https://github.com/cross-language-cpp/djinni-generator/blob/5bb1846ac4db44aa53a4720623f1465214522a90/src/it/scala/djinni/GeneratorIntegrationTest.scala#L333), it doesn't generate `#include "nn.hpp"`. But if we set `objc = true`, `java = true`, and `cppCLI = true`, we will see the issue:
    ```c++
    // jni-headers/dependent_interface.hpp
    #include nn.hpp

    // objcpp-headers/ITDependentInterface+Private.h
    #include nn.hpp

    // cppcli/DependentInterface.hpp
    // It doesn't generate #include nn.hpp in the current `main` branch.
    ```
3. In my branch, this issue happens because I added a test generating C++/CLI code with `useNNHeader = true`.

I will change the generator for all C++, JNI, Objective-C++, and C++/CLI in my branch, but let's deal with the tests separately:

## C++ (`CppMarshal.scala`)
```
spec.cppNnHeader match {
    case Some(nnHdr) => ImportRef(q(nnHdr)) :: base
...
```

## JNI (`JNIGenerator.scala`)
```
spec.cppNnHeader match {
    case Some(nnHdr) => jniHpp.add("#include " + q(nnHdr))
...
```

## Objective-C++ (`ObjcppGenerator.scala`)
```
spec.cppNnHeader match {
    case Some(nnHdr) => refs.privHeader.add("#include " + q(nnHdr))
...
```

## C++/CLI (`CppCliGenerator.scala`)
```
spec.cppNnHeader match {
    case Some(nnHdr) =>
        refs.hpp.add("#include " + q(nnHdr))
...
```

Then, if someone who wrote a batch / bash script to generate the code with `--cpp-nn-header "path/to/nn.hpp"` option before, would need to remove double quotes in their script.

[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/main/scala/djinni/CppMarshal.scala:125]: @a4z I apologize that you feel ignored 😓, when I was still typing the comment, your most recent reply was not shown yet... Hope my above explanation helps to understand the issue a bit better and we can find an agreement how to deal with the case

File: src/it/scala/djinni/IntegrationTest.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala]: @jothepro came up with a bit better implementation, please check changes in [this commit](https://github.com/cross-language-cpp/djinni-generator/pull/128/commits/427a746deb5ddde722e6f1e6e54492baa1759661).

Basically I split string / file comparisons into three cases:
1. **Comparing generated header and source files**: read line by line; compare each line from two files are equal.
2. **Comparing `generated-files.txt`**: read line by line; convert each line to path and compare if two paths are equal.
3. **Comparing printed messages when running generator**, e.g. (`output should equal("Parsing...\nResolving...\nGenerating...\n")`): Replace the line separator in `output` string with "\n"

Hope it looks better now.

File: src/it/scala/djinni/GeneratorIntegrationTest.scala (hunk 1 above)
Comment (FanglinIfolor) [reply src/it/scala/djinni/GeneratorIntegrationTest.scala]: @jothepro I split the tests:
1. C++/CLI interface with external dependencies
2. C++/CLI interface with circular dependencies
3. C++/CLI interface with non-null pointers

Could you review again the changes?

**Comment by FanglinIfolor** [comment]
I pushed new commits trying to resolve above conversations. I can also merge to corresponding commits if everything looks fine.

File: src/main/scala/djinni/CppMarshal.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (a4z) [reply src/main/scala/djinni/CppMarshal.scala]: No worries, didn't mean it very serious. 
I am sorry if my way of expressing me is clumsy and made you feel you have to apologize, now I have to do so. Sorry!

Thanks you so much for the fix, that looks much better and consistent now! Awesome!!

[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/main/scala/djinni/CppMarshal.scala]: Haha, let me grab my broken heart pieces and glue them again (🤲💔 🪄❤️‍🩹), no worries 😄.

File: src/it/scala/djinni/IntegrationTest.scala [hunk 5]
```diff
@@ -193,12 +213,57 @@ class IntegrationTest extends AnyFunSpec {
...
 
+  /** Asserts that all generated files are the same. It basically compares the
+    * paths listed in generated files
+    * `resources/result/$idl/$lang/generated-files.txt` with the expectations
+    * defined in `resources/expected/$idl/$lang/generated-files.txt`.
+    * @param idl
+    *   filename of the input-idl (without file extension)
+    * @param lang
+    *   language to assert for (e.g. `cpp`, `cpp-headers`, `java`, `jni`,
+    *   `jni-headers`, `objc`, `objc-headers`, `objcpp`, `objcpp-headers`)
```
Comment (jothepro) [inline src/it/scala/djinni/IntegrationTest.scala:241]: Copy and paste error? There is no `lang` parameter in `assertGeneratedFilesEqual`.

```suggestion
```

File: src/it/resources/expected/cppcli_circular_dependent_interface/cppcli/AnotherInterface.hpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
+// AUTOGENERATED FILE - DO NOT MODIFY!
+// This file was generated by Djinni from cppcli_circular_dependent_interface.djinni
+
+#pragma once
+
+#include "../cpp-headers/another_interface.hpp"
+#include "OneInterface.hpp"
+#include <memory>
+
+ref class OneInterface;
```
Comment (jothepro) [inline src/it/resources/expected/cppcli_circular_dependent_interface/cppcli/AnotherInterface.hpp:10]: I'm confused, I don't get how just the forward-declaration of the type alone resolves the problem with the circular dependency. Afaik the dependency resolution in this case should still face be the same recursive problem?

`#include "OneInterface.hpp"` -> contains `#include "AnotherInterface.hpp"` -> contains `#include "OneInterface.hpp"`, ...

As far as I understand it to break the circle we would really have to detect that there is a circle and not include the other file in one of the headers? Correct me if I'm wrong though! 🤔

File: src/it/scala/djinni/GeneratorIntegrationTest.scala (hunk 1 above)
Comment (jothepro) [reply src/it/scala/djinni/GeneratorIntegrationTest.scala]: Nice! thx!

File: src/it/scala/djinni/IntegrationTest.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jothepro) [reply src/it/scala/djinni/IntegrationTest.scala]: > Comparing printed messages when running generator [...] Replace the line separator in output string with "\n"

Oh wow, that's nasty! :D 

Off-topic: I wonder if we should automatically build & test the generator on a Windows runner as well to ensure that future changes don't break the tests again. 🙈 But I think that's out of the scope of this PR. 

> Hope it looks better now.

I also feel like your new approach, while being a lot more verbose that the "hack" that you first implemented, feels like a cleaner solution. But only time will tell I guess... Thanks for looking into this again!

**Review by jothepro — COMMENTED** [review]
Thx!

>  I can also merge to corresponding commits if everything looks fine.

As this PR contains multiple unrelated changes, that sounds like a good idea! Normally we would just squash and merge, but in this case a rebase on main may be a good idea. 👍

File: src/it/resources/expected/cppcli_circular_dependent_interface/cppcli/AnotherInterface.hpp (hunk 6 above)
Comment (FanglinIfolor) [reply src/it/resources/expected/cppcli_circular_dependent_interface/cppcli/AnotherInterface.hpp:10]: I guess `#pragma once` in each header does the trick. It seems there are many discussions, e.g. http://www.cplusplus.com/forum/general/243162/.

File: src/it/scala/djinni/IntegrationTest.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala]: > Off-topic: I wonder if we should automatically build & test the generator on a Windows runner as well to ensure that future changes don't break the tests again. 🙈 But I think that's out of the scope of this PR.

I think it's helpful if we would like to support developing the generator on Windows properly, but not sure how many people would actually do it on Windows, given the fact that the test is probably broken for a while on Windows, but no one complained so far.

> I also feel like your new approach, while being a lot more verbose that the "hack" that you first implemented, feels like a cleaner solution. But only time will tell I guess... Thanks for looking into this again!

Any better ideas for cleaner solution would be very welcome 😅

[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala]: Not really much better idea than before, but would still work with only one `assertFileContentEquals` function like before:

```scala
  /** Asserts that all expected files have been created & have the expected
    * content. It basically compares the contents of the generator output in
    * `resources/result/$idl/$lang/$filename` with the expectations defined in
    * `resources/expected/$idl/$lang/$filename`.
    * @param idl
    *   filename of the input-idl (without file extension)
    * @param lang
    *   language to assert for (e.g. `cpp`, `cpp-headers`, `java`, `jni`,
    *   `jni-headers`, `objc`, `objc-headers`, `objcpp`, `objcpp-headers`)
    * @param filenames
    *   list of expected filenames that should have been generated for the given
    *   language
    * @param lineTransformation
    *   function to transform a line, e.g. to a Path type
    */
  def assertFileContentEquals(
      idl: String,
      lang: String,
      filenames: List[String],
      lineTransformation: String => Any = (s: String) => s
  ): Unit = {
    for (filename <- filenames) {
      val resultFile =
        Source.fromFile(s"src/it/resources/result/$idl/$lang/$filename")
      val expectationFile =
        Source.fromFile(s"src/it/resources/expected/$idl/$lang/$filename")
      val resultFileLines = resultFile.getLines
      val expectationFileLines = expectationFile.getLines
      for (
        (result, expectation) <- resultFileLines.zipAll(
          expectationFileLines,
          "",
          ""
        )
      ) {
        lineTransformation(result) should equal(lineTransformation(expectation))
      }

      resultFile.close()
      expectationFile.close()
    }
  }
```

Then, for comparing generated-files.txt, we can use this function like:
```
assertFileContentEquals(idlFile, "", List("generated-files.txt"), s => Paths.get(s))
```

File: src/it/scala/djinni/IntegrationTest.scala (hunk 5 above)
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala:241]: I'll remove it when I push reorganized commits; I would like to wait until we resolve above conversations :)

File: src/it/scala/djinni/IntegrationTest.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (jothepro) [reply src/it/scala/djinni/IntegrationTest.scala]: Sry for my late response! I didn't see that comment.

Why is `lineTransformation` exposed as a parameter? What would be use cases for another line transformation function?

Seeing all the overhead that a clean solution would require made me insecure on what to prefer, the basic string replacement was so... stupid simple... in comparison. 🙈  
Anyways... It seems to me like you invested quite some time to come up with sth better so at this point whatever you propose I'll trust you that it's a good solution! 💪

Afaik this is the only open discussion, is it? I'll wait for you to make a decision on this and then clean up the git history, then we can give the final approval! 👍

[the remark this answers is not in the document]
Comment (FanglinIfolor) [reply src/it/scala/djinni/IntegrationTest.scala]: `lineTransformation` is used specifically for the path comparisons to compare `generated-files.txt` by transforming lines into paths.

Then I keep this version of `assertFileContentEquals`. I have rearranged commits a bit and leave "only" three commits. Hope that's fine and we can merge soon :)

**Comment by a4z** [comment]
Hm, I still need to approve CI runs ... seems the are remembered on a per/repo base.

@FanglinIfolor , I send you a developer invitation to the organisation, please accept, hopefully that should give you instant CI rund

**Review by jothepro — APPROVED** [review]
Nice, thank you for you efforts and patience!

**Review by a4z — APPROVED** [review]
Wonderful to see the C# port got some more ❤️ !

I can just repeat what Johannes said, thank you for you efforts and patience!
